datasets/__init__.py

```diff
@@ -21,6 +25,8 @@ def convert(self, value, param, ctx):
             return DatasetPlugin.from_keys(context=DatasetPlugin._executor.context, **params)
         elif isinstance(value, dict):
             return DatasetPlugin.from_keys(context=DatasetPlugin._executor.context, **value)
+        elif dataclasses.is_dataclass(value):
+            return DatasetPlugin.from_keys(context=DatasetPlugin._executor.context, **asdict(value))
```

Comment:
Looks like the `from_keys` logic tries to find the best-matched dataset plugin type based on param set intersection, which makes sense if we pass in dict/str because they don't have any type hint. But with dataclass do we think maybe we can simply store the dataclass<>dataset plugin mapping relationship somehow (like register the dataclass along with the plugin class) and simply use the mapping and type of dataclass to infer type of dataset plugin to be used (instead of looping through all plugins to find the best fix). I guess that would be more extensible as we add more and more plugins (either internal teams or OSS contributions)? Specifically when designing a new plugin we don't have to manually calculate the intersections in case there's a tie (sometimes the plugin developer might be even not aware of a full list of existing plugins to check the compatibility). Short term might be fine either way though as we only have limited plugins.